Fixes ReconcileVA for non-migratable drivers

yuga711 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

This PR fixes an issue in the volume attachment reconciler. Currently, the reconciler calls a method (unconditionally) that is supposed to be called only when the corresponding volume driver implements the in-tree plugin. 
The PR also fixes a few existing broken tests.

**Which issue(s) this PR fixes**:

Fixes #243

**Special notes for your reviewer**:

Tested the fix with csi-mock driver. The driver does not implement in-tree plugin.
Without the fix, the reconciler failed as below:

```
W0806 02:35:26.595059       1 csi_handler.go:159] Failed to repair volume handle  for driver io.kubernetes.storage.mock: could not find In-Tree driver name for CSI plugin io.kubernetes.storage.mock
```

With the fix, no such failure was observed.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fixes an issue that results in a failure in volume attachment reconciler when the corresponding driver does not implement in-tree plugin. 
```

yuga711 commented on pkg/controller/csi_handler.go:
```diff
@@ -354,6 +357,22 @@ func (h *csiHandler) hasVAFinalizer(va *storage.VolumeAttachment) bool {
+		}
+		return h.translator.IsPVMigratable(pv)
+	} else if va.Spec.Source.InlineVolumeSpec != nil {
+		return va.Spec.Source.InlineVolumeSpec.CSI != nil
```
@msau42: I am not sure if this is the right check for inline vols. Please suggest.

jsafrane replied:
It looks like the right check.

jsafrane commented:
/ok-to-test

jsafrane commented on vendor/modules.txt:
```diff
@@ -3,10 +3,13 @@ github.com/beorn7/perks/quantile
 # github.com/blang/semver v3.5.0+incompatible
 github.com/blang/semver
 # github.com/container-storage-interface/spec v1.2.0
+## explicit
```
You must use golang 1.13 to generate vendor. And why is there any vendor change at all? There is no new import in the code.

yuga711 replied:
Yes, fixed it.

jsafrane commented:
Can you please check the unit tests? They test with "com.test.foo" driver. Why did they succeed? Should they be fixed? Or can you  please add a new unit test that covers your code change?

msau42 commented on pkg/controller/csi_handler.go:
```diff
@@ -354,6 +357,22 @@ func (h *csiHandler) hasVAFinalizer(va *storage.VolumeAttachment) bool {
+	if va.Spec.Source.PersistentVolumeName != nil {
+		pv, err := h.pvLister.Get(*va.Spec.Source.PersistentVolumeName)
+		if err != nil {
+			return false
```
should we return error instead of false?

yuga711 replied:
Yes, makesense.

msau42 commented:
Can you update the PR title and description to be more specific to the issue, something like "Fixes ReconcileVA for non-migratable drivers"

yuga711 commented:
> > Can you please check the unit tests? They test with "com.test.foo" driver. Why did they succeed? Should they be fixed? Or can you please add a new unit test that covers your code change?
> 
> I was curious about this too. We seem to already have unit tests for [ReconcileVA](https://github.com/kubernetes-csi/external-attacher/blob/298cb965f909f76f2d2dfa2062cb5e142f14b9d5/pkg/controller/csi_handler_test.go). I think it's because the fake for RepairVolumeHandle [always succeeds](https://github.com/kubernetes-csi/external-attacher/blob/298cb965f909f76f2d2dfa2062cb5e142f14b9d5/pkg/controller/framework_test.go#L528). I wonder though, why do we need a fake translator? Can we use the real one?

Switched to real translator. Added new checks. Fixed several broken existing tests.

msau42 commented on pkg/controller/csi_handler.go:
```diff
@@ -153,11 +153,19 @@ func (h *csiHandler) ReconcileVA() error {
+		isMig, err := h.isMigratable(va)
 		if err != nil {
-			klog.Warningf("Failed to repair volume handle %s for driver %s: %v", volumeHandle, source.Driver, err)
+			klog.Warningf("Failed to check if migratable: %v", err)
```
Can you log the volume handle too?

yuga711 replied:
ack

yuga711 replied:
Done

msau42 commented on pkg/controller/csi_handler.go:
```diff
@@ -354,6 +362,22 @@ func (h *csiHandler) hasVAFinalizer(va *storage.VolumeAttachment) bool {
+		}
+		return h.translator.IsPVMigratable(pv), nil
+	} else if va.Spec.Source.InlineVolumeSpec != nil {
+		return va.Spec.Source.InlineVolumeSpec.CSI != nil, nil
```
I think it should be an error if InlineVolumeSpec is set, but CSI is nil

yuga711 replied:
ack

yuga711 replied:
Done

msau42 commented on pkg/controller/framework_test.go:
```diff
@@ -297,6 +302,13 @@ func runTests(t *testing.T, handlerFactory handlerFactory, tests []testCase) {
 			test.additionalCheck(t, test)
 		}
+		// makesure all the csi calls were executed.
+		if csiConnection.index < len(csiConnection.calls) {
```
Is it guaranteed that the test will run the reconciler loop exactly the same number of times?

yuga711 replied:
I think yes. Otherwise, I think we need to fix such tests.

msau42 commented on pkg/controller/csi_handler_test.go:
```diff
@@ -34,6 +34,7 @@ import (
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
 	core "k8s.io/client-go/testing"
+	csitranslator "k8s.io/csi-translation-lib"
```
Can we add a test case for ReconcileVA using a volume type that does support migration?

yuga711 replied:
[These](https://github.com/kubernetes/kubernetes/blob/ded1f58779d0d2570735a6ebe7428d57a100ca97/staging/src/k8s.io/csi-translation-lib/translate.go#L29) are the ones that truly support in the actual translator. Do you suggest unit tests based on this? (or) we could setup a fake translator that migration support. If it's more work, can we do this separately?

msau42 replied:
I think adding a unit test using one of the supported translated types would be good.

msau42 commented:
Did you try running the fixed tests without your fix to ReconcileVA, to see if the test fails as expected?

yuga711 commented:
> Did you try running the fixed tests without your fix to ReconcileVA, to see if the test fails as expected?

Yes, the tests fail as expected when I ran them without fix. ie the tests complain saying the expected number of csi calls were not run...

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -696,7 +697,7 @@ func TestCSIHandler(t *testing.T) {
 			name:           "unknown node -> error",
 			initialObjects: []runtime.Object{pvWithFinalizer()},
-			addedVA:        va(false, fin, ann),
+			addedVA:        va(false, fin, nil),
```
Is this change necessary? My tests shows that presence of the annotation does not make any difference, because status is patched, not the object with its finalizers. Having `ann` in the `addedVA` is more correct, as the Patch below expects it's present in the original object.

jsafrane replied:
This applies to all `ann` removed from `addedVA` below

yuga711 replied:
Yes, agreed. Added ann back.

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -833,8 +834,17 @@ func TestCSIHandler(t *testing.T) {
+					types.MergePatchType, patch(va(false /*attached*/, fin /*finalizer*/, ann /* annotations */),
+						vaWithAttachError(va(false, fin, ann), "context deadline exceeded")),
+					"status"),
+				core.NewPatchAction(vaGroupResourceVersion, metav1.NamespaceNone, testPVName+"-"+testNodeName,
```
Please add a comment that the finalizer is added again as a workaround / result of https://github.com/kubernetes-csi/external-attacher/issues/228

yuga711 replied:
done

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -1381,19 +1395,33 @@ func TestCSIHandler(t *testing.T) {
 				// Intentionally empty
 			},
+			expectedActions: []core.Action{
+				// Finalizer is saved first
+				core.NewPatchAction(vaGroupResourceVersion, metav1.NamespaceNone, testPVName+"-"+testNodeName,
```
I think that saving of finalizers has been tested thoroughly in `TestCSIHandler` and here we don't need to test it again. So feel free to set it in `initialObjects`.

yuga711 replied:
done

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -1381,19 +1395,33 @@ func TestCSIHandler(t *testing.T) {
+						va(false /*attached*/, fin, nil))),
+				core.NewPatchSubresourceAction(vaGroupResourceVersion, metav1.NamespaceNone, testPVName+"-"+testNodeName,
+					types.MergePatchType, patch(va(false /*attached*/, "", nil),
+						va(false, "", nil)), "status"),
```
This change is odd:

* The patch is empty.  My tests show that the attacher indeed executes empty PATCH `{}`. Is there a bug in the attacher?

* While the resulting patch is empty, the way it's constructed is wrong - it expects that the volume gets detached (`va(false, ...`) - should it be `true`? The volume was attached successfully.

yuga711 replied:
Haven't found if there is a bug executing empty patch.
Fixed the patch construction logic in the test.

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -1414,11 +1442,12 @@ func TestCSIHandlerReconcileVA(t *testing.T) {
 			initialObjects: []runtime.Object{
-				va(false /*attached*/, "" /*finalizer*/, nID /*annotations*/),
+				deleted(va(false /*attached*/, "" /*finalizer*/, nID /*annotations*/)),
+				// va(false /*attached*/, "" /*finalizer*/, nID /*annotations*/),
```
good catch, please remove the commented line

yuga711 replied:
thanks. done.

yuga711 commented on pkg/controller/framework_test.go:
```diff
@@ -224,6 +224,11 @@ func runTests(t *testing.T, handlerFactory handlerFactory, tests []testCase) {
 					t.Errorf("Failed to reconcile Volume Attachment objects: %v", err)
 				}
 			}
+			if !firstIter && (ctrl.vaQueue.Len() > 0 || ctrl.pvQueue.Len() > 0) {
```
You are right. Added the iteration logic while experimenting with the tests.

jsafrane commented on pkg/controller/csi_handler_test.go:
```diff
@@ -1381,31 +1396,56 @@ func TestCSIHandler(t *testing.T) {
 				{"attach", testVolumeHandle, testNodeID, nil, nil, false, nil, false, nil, 0},
 			},
 		},
+		/* TBD
+		{
+			name:           "GCE PD va attached actual state not attached",
```
In general, the patch looks good, is this the only thing missing?

yuga711 replied:
Here, I wanted to add a test with one of the supported translator types. Now planning to do that separately. I have created a github issue and added a TODO here.

yuga711 replied:
> In general, the patch looks good, is this the only thing missing?
I wanted to add a test with one of the supported translator types. Now planning to do that separately. I have created a github issue and added a TODO here.

jsafrane commented:
/lgtm
/approve
